main/Repl.hs

```diff
@@ -51,40 +55,134 @@ import           Control.Monad.Identity
 import           Control.Monad.Reader
 import           Control.Monad.State.Strict
 
-import           System.Console.Repline        hiding ( options, prefix )
-import           System.Environment
-import           System.Exit
+import           System.Console.Haskeline.Completion
+                                                ( Completion(isFinished)
+                                                , completeWordWithPrev
+                                                , simpleCompletion
+                                                , listFiles
+                                                )
+import           System.Console.Repline         ( Cmd
+                                                , CompletionFunc
+                                                , CompleterStyle (Prefix)
+                                                , ExitDecision(Exit)
+                                                , HaskelineT
+                                                , WordCompleter
+                                                )
+import qualified System.Console.Repline
+import qualified System.Exit
+import qualified System.IO.Error
+
+-- | Repl entry point
+main :: (MonadNix e t f m, MonadIO m, MonadMask m) =>  m ()
+main = main' Nothing
+
+-- | Principled version allowing to pass initial value for context.
+--
+-- Passed value is stored in context with "input" key.
+main' :: (MonadNix e t f m, MonadIO m, MonadMask m) => Maybe (NValue t f m) -> m ()
+main' iniVal = initState iniVal >>= \s -> flip evalStateT s
+    $ System.Console.Repline.evalRepl
+        banner
+        cmd
+        options
+        (Just commandPrefix)
+        (Just "paste")
+        completion
+        (rcFile >> greeter)
+        finalizer
+ where
+  commandPrefix = ':'
 
+  banner = pure . \case
+    System.Console.Repline.SingleLine -> "hnix> "
+    System.Console.Repline.MultiLine  -> "| "
 
-main :: (MonadNix e t f m, MonadIO m, MonadMask m) => m ()
-main = flip evalStateT initState
-    $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
- where
-  prefix = "hnix> "
-  welcomeText =
+  greeter =
     liftIO
       $  putStrLn
       $  "Welcome to hnix "
       <> showVersion version
       <> ". For help type :help\n"
+  finalizer = do
+    liftIO $ putStrLn "Goodbye."
+    return Exit
+
+  rcFile = do
+    f <- liftIO $ Data.Text.IO.readFile ".hnixrc" `catch` handleMissing
+    forM_ (map (words . Data.Text.unpack) $ Data.Text.lines f) $ \case
+      ((prefix:command) : xs) | prefix == commandPrefix -> do
+        let arguments = unwords xs
+        optMatcher command options arguments
+      x -> cmd $ unwords x
+
+  handleMissing e
+    | System.IO.Error.isDoesNotExistError e = return ""
+    | otherwise = throwIO e
+
+  -- Replicated and slightly adjusted `optMatcher` from `System.Console.Repline`
```

Comment:
Could you include a description of the adjustment here?

Reply:
Done, added a TODO for myself to file a PR exposing this from `repline` or adding support for `.rc` directly.